File: ocaml/lib/stack.ml
type 'a node = { value: 'a; next: 'a node option }

type 'a t = {
  mutable head: 'a node option;
  mutable size : int;
}
let create () =
  { head = None; size = 0 }

let push stack value =
  stack.size <- stack.size + 1;
  let new_node = Some({ value; next = stack.head }) in
  stack.head <- new_node

let pop stack =
  match stack.head with
    | None -> None
    | Some(node) ->
      stack.size <- stack.size - 1;
      stack.head <- node.next;
      Some(node.value)

let peek stack =
  match stack.head with
    | None -> None
    | Some(node) -> Some(node.value)

let size stack = stack.size

let to_list stack = 
  let rec append list node =
    match node.next with
    | None -> node.value :: list
    | Some(next) -> node.value :: append list next in
  match stack.head with
  | None -> []
  | Some(node) -> append [] node

File: go/dsa/stack.go
package dsa

import (
	"fmt"
)

type Node[T any] struct {
	Value T
	Next  *Node[T]
}

type Stack[T any] struct {
	Head *Node[T]
}

func NewStack[T any]() *Stack[T] {
	return &Stack[T]{Head: nil}
}

func (s *Stack[T]) Push(value T) {
	newNode := &Node[T]{Value: value, Next: s.Head}
	s.Head = newNode
}

func (s *Stack[T]) Pop() (T, bool) {
	if s.Head == nil {
		var empty T
		return empty, false
	}
	value := s.Head.Value
	s.Head = s.Head.Next
	return value, true
}

func (s *Stack[t]) Print() {
	current := s.Head
	for current != nil {
		fmt.Printf("%v ", current.Value)
		current = current.Next
	}
	fmt.Println()
}
